torch/utils/flop_counter.py

```diff
@@ -0,0 +1,306 @@
+
+            @staticmethod
+            def backward(ctx, *grad_outs):
+                self.parents.append(name)
+                return grad_outs
+
+        return PushState.apply
+
+    def _create_backwards_pop(self, name):
+        class PopState(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, *args):
+                args = tree_map(lambda x: x.clone() if isinstance(x, torch.Tensor) else x, args)
+                if len(args) == 1:
+                    return args[0]
+                return args
+
+            @staticmethod
+            def backward(ctx, *grad_outs):
+                assert(self.parents[-1] == name)
+                self.parents.pop()
+                return grad_outs
+
+        return PopState.apply
+
+    def register_hierarchy(self, f, name):
+        @functools.wraps(f)
+        def wrapped_f(*args, **kwargs):
+            self.parents.append(name)
+            args = self._create_backwards_pop(name)(*args)
+            out = f(*args, **kwargs)
+            assert(self.parents[-1] == name)
+            self.parents.pop()
+            out = normalize_tuple(out)
+            out = self._create_backwards_push(name)(*out)
+            return out
+        return wrapped_f
+
+    def get_table(self, depth=None):
+        if depth is None:
+            depth = self.depth
+        if depth is None:
+            depth = 999999
+        import tabulate
+        tabulate.PRESERVE_WHITESPACE = True
+        header = ["Module", "FLOPS"]
+        values = []
+
+        for mod in self.flop_counts.keys():
+            total_flops = sum(self.flop_counts[mod].values())
+            mod_str = mod
+            mod_depth = mod.count(".")
+
+            if mod_str == "":
+                continue
+            elif mod_str != "Global":
+                mod_depth += 1
+            if mod_depth > depth:
+                continue
+
+            padding = " " * mod_str.count(".")
+            values.append([padding + mod_str, total_flops])
+            for k, v in self.flop_counts[mod].items():
+                values.append([padding + " - " + str(k), v])
+        if len(values) == 0:
+            values = [["Global", 0]]
+        for value in values:
+            value[1] = convert_num_to_suffix_str(value[1])
+        return tabulate.tabulate(values, headers=header, colalign=("left", "right"))
+
+    def __enter__(self):
+        self.flop_counts.clear()
+        super().__enter__()
+
+    def __exit__(self, *args):
+        if self.display:
+            print(self.get_table(self.depth))
+        super().__exit__(*args)
+
+    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        kwargs = kwargs if kwargs else {}
+        out = func(*args, **kwargs)
+        func_packet = func._overloadpacket
+        if func_packet in self.flop_mapping:
+            flop_count_func = self.flop_mapping[func_packet]
+            args_shape, out_shape = tree_map(get_shape, (args, normalize_tuple(out)))
```

Comment:
Non-tensor args for addmm/baddbmm could throw off flop_count functions I think

Reply:
Since if they're 0 we don't actually do the computation?

Reply:
But `get_shape` would still return them unchanged, and then `addmm_flops` will have 5 args instead of 3?

Reply:
I think those are always passed in as kw-only args.

Reply:
Ah cool, yeah docs say kw-only